tests/host_reference/clover_reference.cpp

```diff
@@ -351,6 +411,207 @@ void tmc_matpc(void *out, void **gauge, void *in, void *clover, void *cInv, doub
+                        QUDA_TWIST_GAMMA5_DIRECT, precision);
+
+  // combine
+  xpay(tmpOdd1, -kappa, outOdd1, Vh * spinor_site_size, precision);
+  xpay(tmpOdd2, -kappa, outOdd2, Vh * spinor_site_size, precision);
+
+  xpay(tmpEven1, -kappa, outEven1, Vh * spinor_site_size, precision);
+  xpay(tmpEven2, -kappa, outEven2, Vh * spinor_site_size, precision);
+
+  free(tmpOdd1);
+  free(tmpOdd2);
+  //
+  free(tmpEven1);
+  free(tmpEven2);
+}
+
+// daggerBit && (QUDA_MATPC_EVEN_EVEN_ASYMMETRIC || QUDA_MATPC_ODD_ODD_ASYMMETRIC)
+//
+//   M_{ee}^{-1}^\dagger (D_{eo})^\dagger M_{oo}^{-1}^\dagger  (oddBit == 0)
+//   M_{oo}^{-1}^\dagger (D_{oe})^\dagger M_{ee}^{-1}^\dagger  (oddBit == 1)
+//
+// daggerBit && (QUDA_MATPC_EVEN_EVEN || QUDA_MATPC_ODD_ODD)
+//
+//   (D_{oe})^\dagger M_{oo}^{-1}^\dagger (oddBit == 0)
+//   (D_{eo})^\dagger M_{ee}^{-1}^\dagger (oddBit == 1)
+//
+// !daggerBit
+//
+//   M_{ee}^{-1} D_{eo} (oddBit == 0)
+//   M_{oo}^{-1} D_{oe} (oddBit == 1)
+//
+void tmc_ndeg_dslash(void *out, void **gauge, void *in, void *clover, void *cInv, double kappa, double mu,
+                     double epsilon, int oddBit, QudaMatPCType matpc_type, int daggerBit, QudaPrecision precision,
+                     QudaGaugeParam &gauge_param)
+{
+  // V-4d volume and Vh=V/2, see tests/utils/host_utils.cpp -> setDims()
+  void *in1 = in;
+  void *in2 = (char *)in1 + precision * Vh * spinor_site_size;
+
+  void *out1 = out;
+  void *out2 = (char *)out1 + precision * Vh * spinor_site_size;
+
+  void *tmp1 = malloc(Vh * spinor_site_size * precision);
+  void *tmp2 = malloc(Vh * spinor_site_size * precision);
+
+  if (daggerBit) {
+    ndegTwistCloverGamma5(tmp1, tmp2, in1, in2, clover, cInv, daggerBit, kappa, mu, epsilon, 1 - oddBit,
+                          QUDA_TWIST_GAMMA5_INVERSE, precision);
+    if (matpc_type == QUDA_MATPC_EVEN_EVEN_ASYMMETRIC || matpc_type == QUDA_MATPC_ODD_ODD_ASYMMETRIC) {
+      void *tmptmp1 = malloc(Vh * spinor_site_size * precision);
+      void *tmptmp2 = malloc(Vh * spinor_site_size * precision);
+      wil_dslash(tmptmp1, gauge, tmp1, oddBit, daggerBit, precision, gauge_param);
+      wil_dslash(tmptmp2, gauge, tmp2, oddBit, daggerBit, precision, gauge_param);
+      ndegTwistCloverGamma5(out1, out2, tmptmp1, tmptmp2, clover, cInv, daggerBit, kappa, mu, epsilon, oddBit,
+                            QUDA_TWIST_GAMMA5_INVERSE, precision);
+      free(tmptmp1);
+      free(tmptmp2);
+    } else {
+      wil_dslash(out1, gauge, tmp1, oddBit, daggerBit, precision, gauge_param);
+      wil_dslash(out2, gauge, tmp2, oddBit, daggerBit, precision, gauge_param);
+    }
+  } else {
+    wil_dslash(tmp1, gauge, in1, oddBit, daggerBit, precision, gauge_param);
+    wil_dslash(tmp2, gauge, in2, oddBit, daggerBit, precision, gauge_param);
+    ndegTwistCloverGamma5(out1, out2, tmp1, tmp2, clover, cInv, daggerBit, kappa, mu, epsilon, oddBit,
+                          QUDA_TWIST_GAMMA5_INVERSE, precision);
+  }
+  free(tmp1);
```

Comment:
`host_free` (continue further down in the file)

Reply:
bfa3da9516449da73099c1ad2d9fe869d0d4ea7e